torch/_refs/__init__.py

```diff
@@ -3256,23 +3256,27 @@ def index_fill(
 def index_fill_(
     x: TensorLike, dim: int, index: TensorLike, value: Union[NumberType, TensorLike]
 ):
+    utils.check(
+        index.ndim <= 1,
+        lambda: f"Index should have dimension 1 or 0 (got {index.ndim})",
+    )
     if isinstance(value, TensorLike):
         utils.check(
             value.ndim == 0,
             lambda: "Only supports 0-dimensional value tensor. "  # type: ignore[union-attr]
             f"Got a tensor with {value.ndim} dimensions.",
         )  # type: ignore[arg-type]
-        return x.clone().index_copy_(dim, index, value)
-    dim = utils.canonicalize_dims(x.ndim, dim)
-    utils.check(
-        index.ndim <= 1,
-        lambda: f"Index should have dimension 1 or 0 (got {index.ndim})",
-    )
-    idx = (slice(None),) * dim + (index,)
-    # Treat scalars as elements of \R^1
-    y = x.unsqueeze(0) if x.ndim == 0 else x
-    y[idx] = value  # type: ignore[assignment]
-    return x
+    else:
+        value = torch.scalar_tensor(
+            value, dtype=x.dtype, layout=x.layout, device=x.device  # type: ignore[arg-type]
+        )
+
+    # index_copy does not broadcast on value so we have to do it manually
+    shape = list(x.shape)
+    if x.ndim > 0:
```

Comment:
don't think that's true, 
```
In [14]: index = torch.tensor([0,0,0,0])

In [15]: value = torch.scalar_tensor(4.)

In [16]: shape = list(x.shape)

In [17]: value = value.expand(shape)

In [18]: value
Out[18]: tensor(4.)

In [19]: x.index_copy_(0, index, value)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [19], in <cell line: 1>()
----> 1 x.index_copy_(0, index, value)
```
whereas `index_fill_` works for these inputs.

Reply:
The following works in this branch:
```python
>>> import torch
>>> x = torch.zeros(5)
>>> torch._refs.index_fill_(x, 0, torch.zeros(4, dtype=torch.long), 4)
tensor([4., 0., 0., 0., 0.])
```

Reply:
Ah, I see, it fails when `x = torch.zeros(())`. I'll fix that.

Reply:
Fixed!